File: DS-MADE-EASY/Stack/Stack.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
#include<limits.h>
/*

# Author			: @RAJ009F
# Topic or Type 	: Stack
# Problem Statement	: stack implementation using arrays
# Description		:
# Complexity		:
=======================
#sample output
----------------------

=======================
*/

struct Stack
{
	int size;
	int top;
	int *arr;
};

struct Stack *initStack(int size)
{
	struct Stack *s = (struct Stack *)malloc(sizeof(struct Stack));
	if(!s)
			return NULL;
	s->size = size;
	s->top = -1;
	s->arr = (int *)malloc(sizeof(int)*size);
	
	return s;
	
	
}

int isEmpty(struct Stack *s)
{
	
	return s->top==-1;
	
}
int isFull(struct Stack *s)
{
	return s->top==s->size-1;
	
}

void doubleStack(struct Stack *s)
{
	s->size *=2;
	s->arr=realloc(s->arr, s->size);
	
}

void push(struct Stack *s, int data)
{
	
	if(isFull(s))
	{
		printf("stack overflow\n");
		doubleStack(s);
	}
		s->arr[++s->top] = data;
	
		
}

int pop(struct Stack *s)
{
	if(!isEmpty(s))
		return s->arr[s->top--];
	else
		
		{
			printf("stack underflow\n");
			return -1;
		}
	
}

int Top(struct Stack *s)
{
	if(isEmpty(s))
		return INT_MAX;
	return s->arr[s->top];
}

int main()
	{
		struct Stack *S = initStack(1);
		int i;
		for(i=0;i<10;i++)
		{push(S,i);
		
		printf("[%d]\n",S->arr[S->top]);
			}
			
		for(i=0;i<10;i++)
		{
			int data= pop(S);
			printf("[%d]\n",data);
			
			
		}
	
	return 0;
	}

File: CTCI/CTCI/LinkedList/Stack.java
import java.io.*;
import java.lang.*;
import java.util.*;

// #include <stdio.h>
// #include <stdlib.h>
// #include <string.h>
// #include<stdbool.h>
/*

# Author			: @RAJ009F
# Topic or Type 	: CTCI
# Problem Statement	: Java 
# Description		: 
# Complexity		: 
=======================
#sample output
----------------------

=======================
*/

class Node{
	
	int data;
	Node next;
	Node(int data)
	{
		this.data = data;
		this.next = null;
		
	}
	
	
	
}

class Stack1

{
	Node top;
	
	int  pop()
	{
		if(top !=null)
		{
			int  item = top.data;
			top = top.next;
			return item;
		}
		return -1;
	}
	
	
	void push(int  item)
	{
		if(top!=null)
		{
			Node t  = new Node(item);
			t.next  = top;
			top = t;;
		}
		else
		{
			top = new Node(item);
		}
		
	}
	
	int  peek()
	{
		return top.data;
	}
	
}

public class Stack
{
	public static void main(String args[])
	{
		Stack1 s  =  new Stack1();
		s.push(1);
		s.push(2);
		System.out.println(s.peek());
		System.out.println(s.pop());
		System.out.println(s.pop());System.out.println(s.pop());
		
	}
	
	
}